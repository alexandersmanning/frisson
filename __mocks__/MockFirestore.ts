// import { isEqual } from "lodash";

export class MockFirebaseMethodError extends Error {}

export class MockResult {
  exists: boolean;

  _data: any;

  constructor(exists: boolean, _data: any) {
    this.exists = exists;
    this._data = _data;
  }

  data() {
    return this._data;
  }
}

const testOrder = {
  created_at: 123,
  stripe_id: "st_123",
  email: "bluemangroupie@gmail.com",
  status: "new",
  name: "Aditya TestUser",
  test: false,
  pet_type: "dog",
};

export class MockDocRef {
  options: Options;

  calls: { set: any[]; get: any[]; update: any[] };

  collectionName: string;

  constructor(collectionName: string, options: Options) {
    this.collectionName = collectionName;
    this.options = options;
    this.calls = {
      set: [],
      get: [],
      update: [],
    };
  }

  clearAllMocks() {
    this.calls = {
      set: [],
      get: [],
      update: [],
    };
  }

  // eslint-disable-next-line consistent-return
  async set(data: any) {
    if (this.options.allowedMethods.includes("set")) {
      this.calls.set.push(data);
      return data;
    }
    if (this.options.allowedMethods.includes("logEmailSent")) {
      const validCollections = [
        "email_order_complete_email_sent",
        "email_upload_email_sent",
      ];
      if (
        validCollections.includes(this.collectionName) // &&
        // isEqual(Object.keys(data), ["created_at"])
      ) {
        this.calls.set.push(data);
      } else {
        throw new MockFirebaseMethodError(
          `logEmailSent method is allowed, but you are doing more than logging. Data: ${JSON.stringify(
            data,
          )} collection: ${this.collectionName}`,
        );
      }
    } else {
      throw new MockFirebaseMethodError(
        `Cannot set in MockDocRef. Data: ${JSON.stringify(data)}`,
      );
    }
  }

  async update(data: any) {
    if (this.options.allowedMethods.includes("update")) {
      this.calls.update.push(data);
      return data;
    }
    if (this.options.allowedMethods.includes("updateOrderStatus")) {
      if (true) {
        // isEqual(Object.keys(data), ["status"])) {
        this.calls.update.push(data);
      } else {
        throw new MockFirebaseMethodError(
          `update order status method is allowed, but you are updating more than the order status. Data: ${JSON.stringify(
            data,
          )}`,
        );
      }
      return data;
    }
    throw new MockFirebaseMethodError(
      `Cannot update in MockDocRef. Data: ${JSON.stringify(data)}`,
    );
  }

  async get(params: any = null) {
    if (this.options.allowedMethods.includes("get")) {
      if (this.collectionName === "orders") {
        this.calls.get.push(params);
        return new MockResult(true, testOrder);
      }
      throw new Error(
        `not sure what to get for collection: ${this.collectionName}`,
      );
    } else {
      throw new MockFirebaseMethodError(
        `Cannot get in MockDocRef. Params: ${JSON.stringify(params)}`,
      );
    }
  }
}

export class MockCollection {
  name: string;

  options: Options;

  _doc: MockDocRef;

  constructor(name: string, options: Options) {
    this.name = name;
    this.options = options;
    this._doc = new MockDocRef(this.name, this.options);
  }

  doc(name = "bar") {
    return this._doc;
  }

  clearAllMocks() {
    this._doc.clearAllMocks();
  }
}

type Method = "get" | "set" | "update" | "updateOrderStatus" | "logEmailSent";
type Options = {
  allowedMethods: Method[];
};

export class MockFirestore {
  options: Options;

  _collections: any;

  constructor(options: Options) {
    this.options = options;
    this._collections = {};
  }

  collection(name = "foo") {
    console.log("collection name:!! ", name);
    if (!this._collections[name]) {
      this._collections[name] = new MockCollection(name, this.options);
    }
    return this._collections[name];
  }

  _getDocRef(collectionName: string) {
    return this.collection(collectionName).doc();
  }

  clearAllMocks() {
    Object.values(this._collections).forEach((collection) => {
      (collection as MockCollection).clearAllMocks();
    });
  }
}
