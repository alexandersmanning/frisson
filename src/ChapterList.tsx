import * as t from "./Types";
import React from "react";
import List from "./components/List";
import { Link } from "react-router-dom";
import Button from "./components/Button";
import { XMarkIcon } from "@heroicons/react/24/outline";
function ChapterItem({
  chapter,
  selected,
  onDelete,
}: {
  chapter: t.Chapter;
  selected: boolean;
  onDelete: () => void;
}) {
  const selectedCss = selected ? "bg-listitemhoverSecondary dark:bg-dmlistitemhoverSecondary" : "";
  return (
    <div
      className={`flex py-xs border-b text-slate-300 border-listBorder dark:border-dmlistBorder hover:bg-listitemhoverSecondary dark:hover:bg-dmlistitemhoverSecondary ${selectedCss}`}
    >
      {" "}
      <div className="flex flex-grow">
        <Link to={`/book/${chapter.bookid}/chapter/${chapter.chapterid}`}>
          <div>{chapter.title}</div>
        </Link>
      </div>
      <div
        className="flex flex-none cursor-pointer items-center mr-xs"
        onClick={onDelete}
      >
        <XMarkIcon className="w-4 h-4 text-slate-400" />
      </div>
    </div>
  );
}

export default function ChapterList({
  chapters,
  bookid,
  selectedChapterId,
  onChange,
}: {
  chapters: t.Chapter[];
  bookid: string;
  selectedChapterId: string;
  onChange: () => void;
}) {
  async function deleteChapter(chapterid: string) {
    const res = await fetch(`/api/deleteChapter`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ chapterid }),
    });
    if (!res.ok) {
      console.log(res.statusText);
      return;
    }
    onChange();
  }

  const newChapter = async () => {
    const res = await fetch("/api/newChapter", {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ bookid }),
    });
    if (!res.ok) {
      console.log("error");
      return;
    }
    await onChange();
  };

  const _items = chapters.map((chapter, index) => (
    <li key={chapter.chapterid}>
      <ChapterItem
        chapter={chapter}
        selected={chapter.chapterid === selectedChapterId}
        onDelete={() => deleteChapter(chapter.chapterid)}
      />
    </li>
  ));

  const newChapterButton = (
    <Button className="mb-xs" rounded={true} onClick={newChapter}>
      New Chapter...
    </Button>
  );
  const items = [newChapterButton, ..._items];

  return <List title="Chapters" items={items} className="bg-sidebarSecondary dark:bg-dmsidebarSecondary" />;
}
