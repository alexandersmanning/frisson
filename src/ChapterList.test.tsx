/**
 * @jest-environment jsdom
 */

import renderer from "react-test-renderer";
import ChapterList from "./ChapterList";
import React from "react";
import { chapter1 } from "../__mocks__/mocks";
import { BrowserRouter } from "react-router-dom";

it("shows the ChapterList panel", () => {
  const props = {
    chapters: [chapter1],
    bookid: "book_1",
    selectedChapterId: "chapter_1",
    onChange: () => {},
    saveChapter: () => {},
    closeSidebar: () => {},
    dispatch: () => {},
    canCloseSidebar: true,
  };
  const component = renderer.create(
    <BrowserRouter>
      <ChapterList {...props} />
    </BrowserRouter>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
