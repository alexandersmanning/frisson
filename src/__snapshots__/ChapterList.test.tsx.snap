// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shows the ChapterList panel 1`] = `
<div>
  <div
    className="p-xs border-r border-listBorder dark:border-dmlistBorder h-full w-full bg-sidebarSecondary dark:bg-dmsidebarSecondary false "
    onDragLeave={[Function]}
    onDragOver={[Function]}
    onDrop={[Function]}
  >
    <div
      className="w-full flex pb-xs border-b border-listBorder dark:border-dmlistBorder"
    >
      <button
        className="relative rounded-md inline-flex items-center text-black dark:text-gray-400  hover:bg-gray-50 ring-0 hover:bg-sidebar bg-sidebarSecondary dark:bg-dmsidebarSecondary dark:hover:bg-dmsidebar mr-xs"
        onClick={[Function]}
        type="button"
      >
        <span
          className="sr-only"
        >
          New
        </span>
        <svg
          aria-hidden="true"
          className="w-4 h-4 xl:w-5 xl:h-5"
          fill="none"
          stroke="currentColor"
          strokeWidth={1.5}
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 4.5v15m7.5-7.5h-15"
            strokeLinecap="round"
            strokeLinejoin="round"
          />
        </svg>
      </button>
      <button
        className="relative rounded-md inline-flex items-center text-black dark:text-gray-400  hover:bg-gray-50 ring-0 hover:bg-sidebar bg-sidebarSecondary dark:bg-dmsidebarSecondary dark:hover:bg-dmsidebar"
        onClick={[Function]}
        type="button"
      >
        <span
          className="sr-only"
        >
          Menu
        </span>
        <div
          className="relative flex"
          data-headlessui-state=""
        >
          <button
            aria-expanded={false}
            className="inline-flex items-center gap-x-1 text-sm font-semibold leading-6 text-gray-900"
            data-headlessui-state=""
            id={null}
            onClick={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            type="button"
          >
            <span
              className="sr-only"
            >
              Menu
            </span>
            <svg
              aria-hidden="true"
              className="w-4 h-4 text-slate-400"
              fill="none"
              stroke="currentColor"
              strokeWidth={1.5}
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M6.75 12a.75.75 0 11-1.5 0 .75.75 0 011.5 0zM12.75 12a.75.75 0 11-1.5 0 .75.75 0 011.5 0zM18.75 12a.75.75 0 11-1.5 0 .75.75 0 011.5 0z"
                strokeLinecap="round"
                strokeLinejoin="round"
              />
            </svg>
          </button>
        </div>
      </button>
      <div
        className="flex-grow items-center text-center"
      >
        <h3
          className="text-sm uppercase font-semibold"
        >
          Chapters
        </h3>
      </div>
      <button
        className="relative rounded-md inline-flex items-center text-black dark:text-gray-400  hover:bg-gray-50 ring-0 hover:bg-sidebar bg-sidebarSecondary dark:bg-dmsidebarSecondary dark:hover:bg-dmsidebar"
        onClick={[Function]}
        type="button"
      >
        <span
          className="sr-only"
        >
          Close
        </span>
        <svg
          aria-hidden="true"
          className="w-4 h-4 xl:w-5 xl:h-5"
          fill="none"
          stroke="currentColor"
          strokeWidth={1.5}
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M6 18L18 6M6 6l12 12"
            strokeLinecap="round"
            strokeLinejoin="round"
          />
        </svg>
      </button>
    </div>
    <ul
      className="pt-xs"
    >
      <li
        className=""
      >
        <div
          className="flex text-black dark:text-slate-300 text-sm xl:text-md items-center rounded-md hover:bg-listitemhover hover:dark:bg-dmlistitemhover bg-listitemhover dark:bg-dmlistitemhover"
        >
          <a
            className="flex flex-grow items-center overflow-hidden  py-xs text-ellipsis whitespace-nowrap mr-xs"
            href="/book/book_1/chapter/chapter_1"
            onClick={[Function]}
          >
            <div
              className=""
            >
              <div
                className="px-xs"
              >
                New job
              </div>
            </div>
          </a>
          <div
            className="flex flex-none cursor-pointer items-center mr-xs"
          >
            <div
              className="relative flex"
              data-headlessui-state=""
            >
              <button
                aria-expanded={false}
                className="inline-flex items-center gap-x-1 text-sm font-semibold leading-6 text-gray-900"
                data-headlessui-state=""
                id={null}
                onClick={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                type="button"
              >
                <span
                  className="sr-only"
                >
                  Menu
                </span>
                <svg
                  aria-hidden="true"
                  className="w-4 h-4 text-slate-400"
                  fill="none"
                  stroke="currentColor"
                  strokeWidth={1.5}
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M6.75 12a.75.75 0 11-1.5 0 .75.75 0 011.5 0zM12.75 12a.75.75 0 11-1.5 0 .75.75 0 011.5 0zM18.75 12a.75.75 0 11-1.5 0 .75.75 0 011.5 0z"
                    strokeLinecap="round"
                    strokeLinejoin="round"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
      </li>
    </ul>
  </div>
</div>
`;
